packages/neon_framework/lib/src/utils/request_manager.dart

```diff
@@ -219,26 +219,46 @@ class RequestManager {
       subject.add(Result.loading());
     }
 
-    final cachedParameters = await _cache?.getParameters(account, cacheKey);
-
-    if (cachedParameters != null) {
-      if (cachedParameters.expires != null && !cachedParameters.isExpired) {
-        final cachedValue = await _cache?.get(account, cacheKey);
-        if (cachedValue != null) {
-          if (!subject.isClosed) {
-            subject.add(Result(unwrap(deserialize(cachedValue)), null, isLoading: false, isCached: true));
-          }
-          return;
-        }
-      }
+    // Fetch the cached data with the cache parameters.
+    final cached = await _cache?.get(account, cacheKey);
+    if (subject.isClosed) {
+      return;
+    }
 
-      if (cachedParameters.etag != null) {
-        CacheParameters? cacheParameters;
+    if (cached != null) {
+      // If the cached data is not expired unwrap and emit it.
+      // Return as the value is up to date.
+      if (cached.parameters case CacheParameters(isExpired: false)) {
+        final unwrapped = unwrap(deserialize(cached.value));
+        subject.add(Result(unwrapped, null, isLoading: false, isCached: true));
+        return;
+
+        // If the cached data expired and has an etag, try to refresh the cache parameters.
+        // If the etag did not change, unwrap and emit the cached data.
+        // Save the new cache parameters and return.
+      } else if (cached.parameters case CacheParameters(:final etag) when etag != null && getCacheParameters != null) {
         try {
-          cacheParameters = await timeout(
-            () async => getCacheParameters?.call(),
+          final newParameters = await timeout(
+            getCacheParameters.call,
             timeLimit: timeLimit,
           );
```

Comment:
Shouldn't the cached data be emitted here before trying to get the new parameters? In case the new etag is the same as before we just need to return and no emit the cached data a second time.

Reply:
In what state should the cached data be emitted?
to be correct we would need to emit it as loading and if the etag still matches just re emit as done and return.

If you like this change I'll do it in a followup as this PR doesn't do any behavior changes to the current approach (and as mentioned by you is already hard to review on its own).

Reply:
Yeah I think this makes more sense than the current logic. Follow up is fine by me :+1: